lib/quicly.c

```diff
@@ -4444,15 +4724,177 @@ static int handle_ping_frame(quicly_conn_t *conn, struct st_quicly_handle_payloa
+         * that retires the newly received connection ID, unless it has already done so for that sequence number. (19.15)
+         * TODO: "unless ..." part may not be properly addressed here (we may already have sent the RCID frame for this
+         * sequence) */
+        schedule_retire_connection_id(conn, frame.sequence);
+        /* do not install this CID */
+        return 0;
+    }
+    /* First, handle retire_prior_to field.
+     * This order is important as it is possible to receive a NEW_CONNECTION_ID frame
+     * such that it retires active_connection_id_limit CIDs and then installs
+     * one new CID.
+     */
+    if (frame.retire_prior_to > conn->super.peer.largest_retire_prior_to) {
+        if (frame.retire_prior_to > conn->super.peer.cid_sequence) {
+            schedule_retire_connection_id(conn, conn->super.peer.cid_sequence);
+            need_to_replace_cid = 1;
+        }
+
+        /* retire CIDs in the spare pool */
+        for (int i = 0; i < QUICLY_LOCAL_ACTIVE_CONNECTION_ID_LIMIT - 1; i++) {
+            struct st_quicly_spare_cid_t *spare_cid = &conn->super.peer.spare_cids[i];
+            if (!spare_cid->is_active || spare_cid->sequence >= frame.retire_prior_to)
+                continue;
+            schedule_retire_connection_id(conn, spare_cid->sequence);
+            spare_cid->is_active = 0;
+        }
+        conn->super.peer.largest_retire_prior_to = frame.retire_prior_to;
+    }
+
+    /* store new CID as a spare */
+
+    int was_stored = 0;
+    for (int i = 0; i < QUICLY_LOCAL_ACTIVE_CONNECTION_ID_LIMIT - 1; i++) {
+        struct st_quicly_spare_cid_t *spare_cid = &conn->super.peer.spare_cids[i];
+        if (spare_cid->is_active) {
+            /* If an endpoint receives a NEW_CONNECTION_ID frame that repeats a previously issued connection ID with
+             * a different Stateless Reset Token or a different sequence number, or if a sequence number is used for
+             * different connection IDs, the endpoint MAY treat that receipt as a connection error of type PROTOCOL_VIOLATION.
+             * (19.15)
+             */
+            if (cmp_cid(&spare_cid->cid, frame.cid) == 0) {
+                if (spare_cid->sequence == frame.sequence &&
+                    memcmp(spare_cid->stateless_reset_token, frame.stateless_reset_token, QUICLY_STATELESS_RESET_TOKEN_LEN) == 0) {
+                    /* likely a duplicate due to retransmission */
+                    return 0;
+                } else {
+                    /* received a frame that carries conflicting information */
+                    return QUICLY_TRANSPORT_ERROR_PROTOCOL_VIOLATION;
+                }
+            }
+            if (spare_cid->sequence == frame.sequence && cmp_cid(&spare_cid->cid, frame.cid) != 0)
+                return QUICLY_TRANSPORT_ERROR_PROTOCOL_VIOLATION;
+
+            continue;
+        }
+
+        store_spare_cid(spare_cid, &frame);
+        was_stored = 1;
```

Comment:
Now I realize that we have a problem in the design.

Consider the case where we have received NCID(seq=1), used that CID, sent RCID(seq=1), and received an ACK for RCID(seq=1). After that, the peer might send NCID(seq=1. That is legal behavior, see https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#section-13.3-5).

In such case, we'd be storing CID<sub>1</sub> and might end up using it, which we shouldn't be.

IMO, what we should be doing is something like the following:
* When we establish a connection, allocate slots in `spare_cids` for seq=1, seq=2, ..., and also retain the maximum sequence number that the peer is allowed to send at this moment.
* When receiving a NEW_CONNECTION_ID frame, check NCID.sequence against that maximum. If it is not above that maximum, check if there is an entry within the `spare_cids` array with given sequence number. If there is, store the given CID there. If there is no corresponding entry within the array, we know that we've already retired that CID.
* When we retire a CID (or handle RTP), we update `space_cid` and the allowed maximum as appropriate.

Reply:
You're right. Thanks for noticing this.

Reply:
I have no doubt we must deal with scenario like you mentioned above.

Having said that, I'm just wondering under what scenario this (NCID with seq=1 being resent) would happen. Some examples I could think are: late arrival of duplicated packet and malicious/buggy peer. Would you mind elaborating more on this?

Reply:
As pointed out earlier, the specification allows a QUIC endpoint to retransmit the entire payload of a packet being sent previously. Keeping that in mind, let's consider the following case, where the egress path is very lossy:
1. endpoint repeatedly receives NCID(CID1), sends ACKs, but the ACKs get lost
2. endpoint uses CID1 to probe a new path
3. path probing fails
4. endpoint sends RCID(CID1)
5. endpoint receives ACK for RCID(CID1)

The point is that there is no guarantee that one of the ACKs sent in step 1 reaches the peer before the peer sends ACK for RCID(CID1). We might not hit this situation in practice, though I'm not sure if that is guaranteed. I'd note that path probing might fail almost immediately in certain cases, for example when the interface that was used for sending the probe becomes unusable.

I think we can consider a similar situation when RTP is involved.

Reply:
👍 Thanks for a good example!

Reply:
> When receiving a NEW_CONNECTION_ID frame, check NCID.sequence against that maximum. If it is not above that maximum, check if there is an entry within the spare_cids array with given sequence number. If there is, store the given CID there. If there is no corresponding entry within the array, we know that we've already retired that CID.

Regarding detection of retired CID, I think quicly can have a principle that it would use CID in  the order of sequence number (which we currently do). That way the retired CID should have a sequence number smaller than the one currently in use for communication (`conn->super.peer.cid_sequence` in this PR).

Reply:
So the detection would look like
```
is_already_retired = frame.sequence < conn->super.peer.cid_sequence;
```
Let me know if you think of any potential flaw in this assumption.

Reply:
This doesn't prove the entire correctness of my algorithm above but as a note to myself here is what will happen when packet reordering happens. Let's say I'm using CID0 and the peer hasn't issued other CIDs yet. Then the peer issues CID1 and CID2:
```
CID1 (rpt=1)
CID2 (rpt=1)
```
but for whatever reason two frames are delivered out-of-order. So my host would see NCID frames in this order:
```
CID2 (rpt=1)
CID1 (rpt=1)
```
I handle CID2 first, and since RPT=1 is specified, CID0 is retired and CID2 becomes the current CID for communication. After that CID1 arrives, but since CID2 is already in use, CID1 is determined as "already retired" according to the above algorithm, so RCID for CID1 is sent.

This algorithm wastes some CIDs in cases like this but I would assume the probability of hitting such a case would be quite low, so I think it's still justifiable given its simplicity.

Reply:
Though nothing forbids us from using CIDs in the order of their sequence numbers, I am not sure if that is always going to work well.

As an example, https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#section-9.1-2 suggests that the peer might send a probe with a NCID frame to make sure that an endpoint 
has a spare CID. But the specification does not say that the peer should send all unacknowledged NCID frames. My understanding is that this is because the specification assumes that endpoints would be capable of using CIDs in any order.

Reply:
Hmm, okay, then maybe we should not stick to the assumption that CIDs are sent and used in the order of sequence number.

Reply:
Updated the branch. I believe what I have now basically follows your algorithm.